23-ima-sig-signature-verification.md

```diff
@@ -0,0 +1,287 @@
+  registered public keys (RSA, DSA, EC, etc)
+- The verifier will require that each IMA measurement list entry will
+  either be verifiable with a key or can be reconciled against
+  a allowlist or exclude list; every IMA measurement list entry must
+  be 'covered'
+- The tenant client tool will be extended to support registering
+  per-system public keys to be used for signature verification
+- Test cases will be written that read in the example keys and
+  IMA measuement lists and perform signature verification along
+  with alllwlist and exclude list reconciliation
+
+### Notes/Constraints/Caveats (optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above.
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+- Signature verification may be more CPU intensive than verification
+  against a small allowlist of measurements; on the other hand
+  the advantage of signature verification is that it only needs a few
+  keys rather than a possibly huge allowlist
+- Every IMA measurement entry should be covered either using a
+  measurement allowlist reconciliation, be part of an exclude list, or
+  the signature verification has to succeed; the complexity on
+  the user side is to setup a system by signing some immutable files
+  while leaving others to allowlist and putting mutable files
+  into an exclude list
+- A prototype of this extension is being built here:
+  https://github.com/stefanberger/keylime/commits/ima_sig
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal and how do we mitigate.  Think broadly.
+For example, consider both security and how this will impact the larger
+enhancement ecosystem.
+
+How will security be reviewed and by whom?
+-->
+
+TBD
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable.  This may include API specs (though not always
+required) or even code snippets.  If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+To make it easy for a client, the tenant tool will be able to read
+public key files in various format, such as PEM or DER formatted
+public keys, x509 certificates, or even private keys. Only the public
+key parts will be stored in the Keylime database.
+
+The existing exclude list will be used for skipping of mutable files.
+If no allowlist is passed via the tenant tool, only signatures Will be
+verified. All IMA measurement list entries that have signatures must be verifiable
+with one of the provided public keys, otherwise the file must be in the excluded
+list.
+If an allowlist, exclude list, and keys for signature verification are provided
+then each measurement list entry that is not skipped due to the exclude list
+must pass both the signature verification and be in the allowlist. Another
```

Comment:
I can see both options as being useful. Some setups might use signatures for everything that's not excluded. Others might use signatures for things like the kernel (or vendor provided) but allowlists for everything they do themselves on top of that.

Reply:
I tried to enumerate the possible cases here. Every entry must be covered by either signature, allow list, or exclude list. Entries may need to be reconciled by allow list *and* signature list, and maybe that should be the default. One could work with a '*or*' as well. Being able to specify the *and* or *or* on a per entry basis may become a real headache to set up.

Reply:
Yeah, I agree we don't need to support per-entry settings. I think globally in keylime.conf for the verifier is fine.

Reply:
If we make this configurable I think it should be an option for `keylime_tenant` to make it per-system configurable.